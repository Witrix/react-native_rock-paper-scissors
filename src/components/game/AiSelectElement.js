import React, { Component } from 'react';
import {
    View,
    FlatList,
    Text,
    StyleSheet,
    TouchableHighlight,
} from 'react-native';
import AppStyle             from '../../themes/styles';
import AppColor             from '../../themes/colors';
import MyLocalization       from "../../my-lib/MyLocalzation";
import MyRules              from '../../my-lib/MyRules';
import Element from "./Element";

export default class AiSelectElement extends Component {

    constructor(props) {
        super(props);
        this.state = {
            mode: props.mode
        };
    }

    componentDidMount() {
        setTimeout(() => {
            let rules = MyRules.getRules(this.state.mode);
            if (rules && rules.length > 0) {
                let number =  Math.floor(Math.random() * Math.floor(rules.length));
                let selected = rules[number].reference;
                this.setState({selected: selected}, () => {
                    setTimeout(() => {
                        this.onPressElement(this.state.selected)
                    }, 1000)
                })
            }
        }, 1000)
    }

    onPressElement(element) {
        if (this.props.onSelect && typeof this.props.onSelect === 'function') this.props.onSelect(element);
    }

    renderItem(item) {
        let element = item.item;
        return (
            <View style={[style.elementContainer, this.state.selected === element.reference ? {backgroundColor: AppColor.positive} : null]}>
                <Element type={element.reference} mode={this.state.mode}/>
            </View>
        );
    }

    render() {
        let rule = MyRules.getRules(this.state.mode);
        return (
            <View style={style.container}>
                <Text style={AppStyle.h2}>{MyLocalization.getString('aiSelectSign')}</Text>
                <FlatList data={rule}
                          keyExtractor={(item, index) => item.reference}
                          numColumns={3}
                          extraData={this.state.selected}
                          columnWrapperStyle={style.container}
                          renderItem={this.renderItem.bind(this)}/>
            </View>
        );
    }
}

const style = StyleSheet.create({
    container: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    elementContainer: {
        padding: 10,
        borderRadius: 10,
    }
});
