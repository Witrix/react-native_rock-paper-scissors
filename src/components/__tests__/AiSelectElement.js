import 'react-native';
import React                    from 'react';
import AiSelectElement          from '../game/AiSelectElement';
import renderer                 from 'react-test-renderer';

it('Render correctly', () => {
  const tree = renderer.create(
    <AiSelectElement />
  );
    expect(tree).toMatchSnapshot();
});

it('Render correctly: valid props', () => {
    const tree = renderer.create(
        <AiSelectElement mode={'classic'}/>
    );
    expect(tree).toMatchSnapshot();
});

it('Render correctly: invalid props', () => {
    const tree = renderer.create(
        <AiSelectElement mode={-1}/>
    );
    expect(tree).toMatchSnapshot();
});