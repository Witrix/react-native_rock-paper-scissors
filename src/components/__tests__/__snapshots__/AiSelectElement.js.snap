// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000000",
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    The AI selects a sign
  </Text>
  <RCTScrollView
    columnWrapperStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    data={
      Array [
        Object {
          "beat": Array [
            "scissors",
          ],
          "beatBy": Array [
            "paper",
          ],
          "localizedKey": "rock",
          "reference": "rock",
          "resourceKey": 1,
        },
        Object {
          "beat": Array [
            "rock",
          ],
          "beatBy": Array [
            "scissors",
          ],
          "localizedKey": "paper",
          "reference": "paper",
          "resourceKey": 1,
        },
        Object {
          "beat": Array [
            "paper",
          ],
          "beatBy": Array [
            "rock",
          ],
          "localizedKey": "scissors",
          "reference": "scissors",
          "resourceKey": 1,
        },
      ]
    }
    disableVirtualization={false}
    extraData={undefined}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Rock
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Paper
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Scissors
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Render correctly: invalid props 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000000",
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    The AI selects a sign
  </Text>
  <RCTScrollView
    columnWrapperStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    data={
      Array [
        Object {
          "beat": Array [
            "scissors",
          ],
          "beatBy": Array [
            "paper",
          ],
          "localizedKey": "rock",
          "reference": "rock",
          "resourceKey": 1,
        },
        Object {
          "beat": Array [
            "rock",
          ],
          "beatBy": Array [
            "scissors",
          ],
          "localizedKey": "paper",
          "reference": "paper",
          "resourceKey": 1,
        },
        Object {
          "beat": Array [
            "paper",
          ],
          "beatBy": Array [
            "rock",
          ],
          "localizedKey": "scissors",
          "reference": "scissors",
          "resourceKey": 1,
        },
      ]
    }
    disableVirtualization={false}
    extraData={undefined}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Rock
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Paper
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Scissors
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Render correctly: valid props 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000000",
        "fontSize": 20,
        "textAlign": "center",
      }
    }
  >
    The AI selects a sign
  </Text>
  <RCTScrollView
    columnWrapperStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    data={
      Array [
        Object {
          "beat": Array [
            "scissors",
          ],
          "beatBy": Array [
            "paper",
          ],
          "localizedKey": "rock",
          "reference": "rock",
          "resourceKey": 1,
        },
        Object {
          "beat": Array [
            "rock",
          ],
          "beatBy": Array [
            "scissors",
          ],
          "localizedKey": "paper",
          "reference": "paper",
          "resourceKey": 1,
        },
        Object {
          "beat": Array [
            "paper",
          ],
          "beatBy": Array [
            "rock",
          ],
          "localizedKey": "scissors",
          "reference": "scissors",
          "resourceKey": 1,
        },
      ]
    }
    disableVirtualization={false}
    extraData={undefined}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Rock
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Paper
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "padding": 10,
                },
                null,
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 70,
                    "width": 70,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Scissors
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
